File: packages/cli/internal/scaffold/scaffold.go
package scaffold

import (
	"bytes"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
	"text/template"
	"time"

	"github.com/agentx-labs/agentx/internal/manifest"
)

// ScaffoldData holds all template variables available to scaffold templates.
type ScaffoldData struct {
	Name        string // e.g., "ssm-lookup"
	Topic       string // e.g., "cloud" (skills only)
	Vendor      string // e.g., "aws" (skills only, may be empty)
	Runtime     string // "node" or "go" (skills only)
	Description string // Human-readable description
	Version     string // Semver, e.g., "0.1.0"
	PackageName string // Derived: @agentx/skill-<topic>-<name>
	SkillPath   string // Derived: <topic>/<vendor>/<name> or <topic>/<name>
	ModuleName  string // Derived: github.com/agentx-labs/agentx-skill-<name>
	Year        int    // Current year
}

// Result holds the outcome of a scaffold generation.
type Result struct {
	OutputDir string
	Files     []string
	Warnings  []string
}

// NewScaffoldData creates a ScaffoldData with derived fields populated.
func NewScaffoldData(name, typeName, topic, vendor, runtime string) *ScaffoldData {
	d := &ScaffoldData{
		Name:    name,
		Topic:   topic,
		Vendor:  vendor,
		Runtime: runtime,
		Version: "0.1.0",
		Year:    time.Now().Year(),
	}

	d.Description = fmt.Sprintf("AgentX %s: %s", typeName, name)

	// Derive skill-specific fields.
	if topic != "" {
		if vendor != "" {
			d.SkillPath = fmt.Sprintf("%s/%s/%s", topic, vendor, name)
			d.PackageName = fmt.Sprintf("@agentx/skill-%s-%s", topic, name)
		} else {
			d.SkillPath = fmt.Sprintf("%s/%s", topic, name)
			d.PackageName = fmt.Sprintf("@agentx/skill-%s-%s", topic, name)
		}
	}

	d.ModuleName = fmt.Sprintf("github.com/agentx-labs/agentx-skill-%s", name)

	return d
}

// templateSetName returns the embedded directory name for a given type+runtime.
func templateSetName(typeName, runtime string) string {
	if typeName == "skill" {
		return "skill-" + runtime
	}
	return typeName
}

// manifestFileName returns the expected manifest file name for a type.
func manifestFileName(typeName string) string {
	return typeName + ".yaml"
}

// Generate creates a new type from scaffolding templates.
func Generate(typeName string, data *ScaffoldData, outputDir string) (*Result, error) {
	setName := templateSetName(typeName, data.Runtime)
	templatesDir := filepath.Join("scaffolds", setName)

	// Verify template set exists in embedded FS.
	entries, err := fs.ReadDir(scaffoldFS, templatesDir)
	if err != nil {
		return nil, fmt.Errorf("template set %q not found: %w", setName, err)
	}

	// Create output directory.
	if err := os.MkdirAll(outputDir, 0755); err != nil {
		return nil, fmt.Errorf("creating output directory: %w", err)
	}

	// Check for existing files to prevent accidental overwrites.
	existingEntries, err := os.ReadDir(outputDir)
	if err == nil && len(existingEntries) > 0 {
		return nil, fmt.Errorf("output directory %s is not empty; remove existing files first", outputDir)
	}

	result := &Result{
		OutputDir: outputDir,
	}

	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}

		tmplPath := filepath.Join(templatesDir, entry.Name())
		tmplBytes, err := fs.ReadFile(scaffoldFS, tmplPath)
		if err != nil {
			return nil, fmt.Errorf("reading template %s: %w", tmplPath, err)
		}

		// Strip .tmpl extension for the output filename.
		outName := strings.TrimSuffix(entry.Name(), ".tmpl")
		outPath := filepath.Join(outputDir, outName)

		// Handlebars (.hbs) files use {{ }} syntax that conflicts with
		// Go's text/template. Copy them verbatim without processing.
		if strings.HasSuffix(outName, ".hbs") {
			if err := os.WriteFile(outPath, tmplBytes, 0644); err != nil {
				return nil, fmt.Errorf("writing %s: %w", outPath, err)
			}
			result.Files = append(result.Files, outName)
			continue
		}

		// Parse and execute the Go template.
		tmpl, err := template.New(entry.Name()).Parse(string(tmplBytes))
		if err != nil {
			return nil, fmt.Errorf("parsing template %s: %w", entry.Name(), err)
		}

		var buf bytes.Buffer
		if err := tmpl.Execute(&buf, data); err != nil {
			return nil, fmt.Errorf("executing template %s: %w", entry.Name(), err)
		}

		if err := os.WriteFile(outPath, buf.Bytes(), 0644); err != nil {
			return nil, fmt.Errorf("writing %s: %w", outPath, err)
		}

		result.Files = append(result.Files, outName)
	}

	// Validate the generated manifest against JSON Schema.
	manifestFile := filepath.Join(outputDir, manifestFileName(typeName))
	if _, err := os.Stat(manifestFile); err == nil {
		valResult, valErr := manifest.ValidateFile(manifestFile)
		if valErr != nil {
			result.Warnings = append(result.Warnings,
				fmt.Sprintf("Could not validate manifest: %v", valErr))
		} else if !valResult.Valid {
			for _, issue := range valResult.Issues {
				msg := issue.Message
				if issue.Path != "" {
					msg = issue.Path + ": " + msg
				}
				result.Warnings = append(result.Warnings, msg)
			}
		}
	}

	return result, nil
}

File: src/core/scaffold.ts
import { join, dirname } from 'node:path';
import {
  readFileSync,
  writeFileSync,
  readdirSync,
  mkdirSync,
  existsSync,
  statSync,
} from 'node:fs';
import { fileURLToPath } from 'node:url';

export interface ScaffoldData {
  name: string;
  topic: string;
  vendor: string;
  runtime: string;
  description: string;
  version: string;
  packageName: string;
  skillPath: string;
  year: number;
}

export interface ScaffoldResult {
  outputDir: string;
  files: string[];
  warnings: string[];
}

export function newScaffoldData(
  name: string,
  typeName: string,
  topic: string,
  vendor: string,
  runtime: string,
): ScaffoldData {
  const skillPath = vendor ? `${topic}/${vendor}/${name}` : `${topic}/${name}`;
  return {
    name,
    topic,
    vendor,
    runtime,
    description: `A ${typeName} for ${name}`,
    version: '0.1.0',
    packageName: `@agentx/${typeName}-${topic}-${name}`,
    skillPath,
    year: new Date().getFullYear(),
  };
}

function templateSetName(typeName: string, runtime: string): string {
  if (typeName === 'skill') return `skill-${runtime}`;
  return typeName;
}

function manifestFileName(typeName: string): string {
  if (['skill', 'workflow', 'prompt', 'persona', 'context', 'template'].includes(typeName)) {
    return `${typeName}.yaml`;
  }
  return 'manifest.yaml';
}

function getScaffoldsDir(): string {
  // At runtime, code runs from dist/ — scaffolds live at src/scaffolds/
  const thisFile = fileURLToPath(import.meta.url);
  const projectRoot = join(dirname(thisFile), '..');
  const candidate = join(projectRoot, 'src', 'scaffolds');
  if (existsSync(candidate)) return candidate;
  throw new Error('Scaffolds directory not found');
}

function renderTemplate(
  template: string,
  data: ScaffoldData,
): string {
  let result = template;
  for (const [key, value] of Object.entries(data)) {
    const pattern = new RegExp(`\\{\\{\\s*\\.${key.charAt(0).toUpperCase() + key.slice(1)}\\s*\\}\\}`, 'g');
    result = result.replace(pattern, String(value));
  }
  return result;
}

export function generate(
  typeName: string,
  data: ScaffoldData,
  outputDir: string,
): ScaffoldResult {
  const setName = templateSetName(typeName, data.runtime);
  const scaffoldsDir = getScaffoldsDir();
  const templateDir = join(scaffoldsDir, setName);

  if (!existsSync(templateDir)) {
    throw new Error(`Template set not found: ${setName}`);
  }

  // Prevent overwriting non-empty directories
  if (existsSync(outputDir)) {
    const entries = readdirSync(outputDir);
    if (entries.length > 0) {
      throw new Error(`Output directory is not empty: ${outputDir}`);
    }
  }

  mkdirSync(outputDir, { recursive: true });
  const files: string[] = [];

  for (const entry of readdirSync(templateDir)) {
    const srcPath = join(templateDir, entry);
    if (!statSync(srcPath).isFile()) continue;

    // Strip .tmpl extension
    const outName = entry.endsWith('.tmpl') ? entry.slice(0, -5) : entry;
    const content = readFileSync(srcPath, 'utf-8');

    // .hbs files are copied verbatim, others are rendered
    const rendered = entry.endsWith('.hbs.tmpl') ? content : renderTemplate(content, data);

    const outPath = join(outputDir, outName);
    writeFileSync(outPath, rendered, 'utf-8');
    files.push(outName);
  }

  return { outputDir, files, warnings: [] };
}
